staging_service/kb_auth_client.py

```diff
@@ -0,0 +1,160 @@
+        # assume that if we get json then at least this is the auth server and we can
+        # rely on the error structure.
+        err = j["error"].get("appcode")
+        if err == 10020:  # Invalid token
+            raise InvalidTokenError("KBase auth server reported token is invalid.")
+        if err == 30010:  # Illegal username
+            # The auth server does some goofy stuff when propagating errors, should be cleaned up
+            # at some point
+            raise InvalidUserError(j["error"]["message"].split(":", 3)[-1])
+        # don't really see any other error codes we need to worry about - maybe disabled?
+        # worry about it later.
+        raise IOError("Error from KBase auth server: " + j["error"]["message"])
+
+
+class KBaseAuth:
+    """A client for contacting the KBase authentication server."""
+
+    @classmethod
+    async def create(
+        cls,
+        auth_url: str,
+        cache_max_size: int = 10000,
+        cache_expiration: int = 300,
+    ) -> Self:
+        """
+        Create the client.
+        auth_url - The root url of the authentication service.
+        cache_max_size -  the maximum size of the token cache.
+        cache_expiration -  the expiration time for the token cache in
+            seconds.
+        """
+        if not _require_string(auth_url, "auth_url").endswith("/"):
+            auth_url += "/"
+        j = await _get(auth_url, {"Accept": "application/json"})
+        return KBaseAuth(
+            auth_url,
+            cache_max_size,
+            cache_expiration,
+            j.get("servicename"),
+        )
+
+    def __init__(
+        self,
+        auth_url: str,
+        cache_max_size: int,
+        cache_expiration: int,
+        service_name: str,
+    ):
+        self._url = auth_url
+        self._me_url = self._url + "api/V2/me"
+        self._cache_timer = time.time  # TODO TEST figure out how to replace the timer to test
+        # cache is intended to map tokens -> usernames
+        self._cache = LRUCache(
+            timer=self._cache_timer, maxsize=cache_max_size, ttl=cache_expiration
+        )
+
+        if service_name != "Authentication Service":
+            raise IOError(
+                f"The service at {self._url} does not appear to be the KBase "
+                + "Authentication Service"
+            )
+
+        # could use the server time to adjust for clock skew, probably not worth the trouble
+
+    async def get_user(self, token: str) -> str:
+        """
+        Get a username from a token.
+
+        token - The user's token.
+
+        Returns the user.
+        Raises a ValueError if the token is missing
+        Raises a InvalidTokenError if the token is invalid
+        """
+        # TODO CODE should check the token for \n etc.
+        _require_string(token, "token")
+
+        if token in self._cache:
+            return self._cache.get(token)
+        result = await _get(self._me_url, {"Authorization": token})
```

Comment:
If you're not getting roles, there's no need to call the me endpoint, the token endpoint will work, puts less load on the db, and is lower bandwidth

Reply:
:+1: